### PR TITLE
LayerScale on Transolver residuals — cross-dataset (TF/TF-paper/AF/DM)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

LayerScale (Touvron et al., CaiT/DeiT-III, 2021) adds a per-layer learnable scalar multiplier α (initialized to a very small ε) to the output of each residual block before the skip connection. At initialization, the network behaves like an identity map, which dramatically stabilises training of deep transformers and often yields better final minima.

For Transolver blocks the residual update is `x = x + Layer(x)`. LayerScale makes it `x = x + α * Layer(x)` where α is a per-channel nn.Parameter vector (shape `[hidden_dim]`) initialized to ε=1e-6 per layer — two α vectors per block: one after self-attention, one after FFN.

This is a standard, low-risk improvement with no hyperparameter sweep needed. It has improved performance on ViT-Large/Huge-class models and on physics simulators that share Transolver's deep-residual structure. It has never been tested in this pipeline.

Reference: https://arxiv.org/abs/2103.17239 (Going Deeper with Image Transformers, §3.1)

## Instructions

Add LayerScale to the Transolver attention block in `target/icml2026/train.py`. Specifically:

1. In the attention block (the class that defines self-attention + FFN residuals), add two `nn.Parameter` vectors per block:
   ```python
   self.ls1 = nn.Parameter(torch.ones(hidden_dim) * 1e-6)  # after attention
   self.ls2 = nn.Parameter(torch.ones(hidden_dim) * 1e-6)  # after FFN
   ```
   Use a per-channel vector (shape `[hidden_dim]`) rather than a scalar — this is the standard CaiT formulation and gives the model more expressivity at negligible parameter cost.

2. Modify the forward pass residual connections:
   ```python
   # Before (current):
   x = x + self.attn(self.norm1(x))
   x = x + self.ffn(self.norm2(x))

   # After (LayerScale):
   x = x + self.ls1 * self.attn(self.norm1(x))
   x = x + self.ls2 * self.ffn(self.norm2(x))
   ```

3. Add a CLI flag `--layer-scale` (boolean, default False) to enable/disable this feature so the baseline is still reproducible.

4. Run all **four datasets** using the golden config for each plus `--layer-scale`. Use `--wandb_group layer-scale-cross-dataset` for all runs.

### TandemFoil run
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --layer-scale \
  --epochs 999 \
  --wandb_group layer-scale-cross-dataset
```

### TandemFoil Paper run
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --layer-scale \
  --epochs 999 \
  --wandb_group layer-scale-cross-dataset
```

### AirfRANS run
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 \
  --ema-decay 0.999 \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --enable-fourier \
  --layer-scale \
  --epochs 999 \
  --wandb_group layer-scale-cross-dataset
```

### DrivAerML run
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier \
  --layer-scale \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --epochs 999 \
  --wandb_group layer-scale-cross-dataset
```

Note: DrivAerML uses `--no-use-ema` (not `--ema-decay 0.999`) — EMA hurts on DM per PR #2899.

## Baseline

Current best validation metrics:

| Dataset | Metric | Best val | Best PR |
|---|---|---|---|
| TandemFoil | `val_primary/surface_pressure_mae` | **26.06** | #2887 |
| TandemFoil Paper | `val_primary/field_mse` | not yet established | — |
| AirfRANS | `val_primary/surface_mse` | **0.000598** | #2906 |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** | #2898 |

Reproduce TandemFoil baseline:
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

Reproduce AirfRANS baseline:
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --no-use-ema \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999
```

Reproduce DrivAerML baseline:
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

## Expected outcome

LayerScale is a stabilising prior on the residual scale that helps optimisers find flatter minima. Training-stability improvements like this tend to reduce the gap between early-epoch and late-epoch performance, which is especially valuable under our fixed epoch budget. Expected: 2–8% improvement across all four datasets, or at minimum no regression.

Report `val_primary/surface_pressure_mae` for TF, `val_primary/field_mse` for TF-paper, `val_primary/surface_mse` for AF, and `val_primary/surface_rel_l2_pct` for DM in your results comment. Also report W&B run IDs for all four runs.

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
